utils/python/CIME/case.py

```diff
@@ -6,49 +6,84 @@
+from CIME.XML.pio                   import PIO
+from CIME.XML.archive               import Archive
+
+from CIME.XML.env_test              import EnvTest
+from CIME.XML.env_mach_specific     import EnvMachSpecific
+from CIME.XML.env_case              import EnvCase
+from CIME.XML.env_mach_pes          import EnvMachPes
+from CIME.XML.env_build             import EnvBuild
+from CIME.XML.env_run               import EnvRun
+from CIME.XML.env_archive           import EnvArchive
+from CIME.XML.env_batch             import EnvBatch
+
+from CIME.XML.generic_xml           import GenericXML
 
 logger = logging.getLogger(__name__)
 
 class Case(object):
 
     def __init__(self, case_root=os.getcwd()):
-        expect(os.path.isdir(case_root),
-               "Case root directory '%s' does not exist" % case_root)
-
-        self._env_entryid_files = []
-        self._env_generic_files = []
 
         self._env_files_that_need_rewrite = set()
 
+        self._env_entryid_files = []
         self._env_entryid_files.append(EnvRun(case_root))
         self._env_entryid_files.append(EnvBuild(case_root))
         self._env_entryid_files.append(EnvMachPes(case_root))
         self._env_entryid_files.append(EnvCase(case_root))
         self._env_entryid_files.append(EnvBatch(case_root))
         if os.path.isfile(os.path.join(case_root,"env_test.xml")):
             self._env_entryid_files.append(EnvTest(case_root))
+        self._env_generic_files = []
         self._env_generic_files.append(EnvMachSpecific(case_root))
         self._env_generic_files.append(EnvArchive(case_root))
+        self._files = self._env_entryid_files + self._env_generic_files
 
-        self._case_root = case_root
+        # Hold arbitary values. In create_newcase we may set values
+        # for xml files that haven't been created yet. We need a place
+        # to store them until we are ready to create the file. At file
+        # creation we get the values for those fields from this lookup
+        # table and then remove the entry. This was what I came up
+        # with in the perl anyway and I think that we still need it here.
+        self.lookups = {}
+        self.lookups['CIMEROOT'] = os.path.abspath(get_cime_root())
+
+        self._compsetname = None
+        self._gridname = None
+        self._compsetsfile = None
+        self._pesfile = None
+        self._gridfile = None
+        self._components = []
+        self._component_config_files = []
 
     def __del__(self):
         self.flush()
 
-    def flush(self):
-        for env_file in self._env_files_that_need_rewrite:
-            env_file.write()
+    def _get_env(self, short_name):
+          full_name = "env_%s.xml" % (short_name)
+          for env_file in self._files:
+              if os.path.basename(env_file.filename) == full_name:
+                  return env_file
+          expect(False, "Could not find object for %s in case"%full_name)
+
+    def flush(self, flushall=False):
```

Comment:
Then there might be a problem with our encapsulation of things. We should require that all updates go through Case.